### ISSUE TITLE
Photo behavior in the cropping area

### ISSUE THREAD

**Issue by vlkonoshenko** [icapps/flutter-custom-image-crop#38]
1. Can make image smaller then crop area.  
2. Can move out image from crop area.
I think its look odd

https://github.com/icapps/flutter-custom-image-crop/assets/15077780/714cc707-fca3-4ca5-855d-92d90e8396f8

**Comment by ikbendewilliam** [comment]
You could listen with the `CustomImageCropController` Calculate if the position is invalid on your own terms and use older data to set the image back inside your bounds with `void setData(CropImageData data)`. 
Just make sure that the `setData` doesn't trigger the listener, since then you would have an endless loop

**Comment by vlkonoshenko** [comment]
@ikbendewilliam  hmm... looks like callbacks not calling now
<img width="1002" alt="Снимок экрана 2023-08-18 в 09 54 49" src="https://github.com/icapps/flutter-custom-image-crop/assets/15077780/68ae2532-6898-4fdf-852e-7657fe3738c4">

**Comment by ikbendewilliam** [comment]
Hmm I'll have to try it out, I'll have a colleague look at it next week 🙂

**Comment by ikbendewilliam** [comment]
@vlkonoshenko I've tried, but I can't get it to work 😅 fully. 

What you'll need to do is add a check in the `onMoveUpdate` and `onScaleUpdate`. Then in the check you'll need to check whether the image is in the bounds. This is the part where I'm having trouble, especially with the rotation. 
This is some code that kind of works (movement and scaling is working). 
But I can't add this to the package unless it works for all customImageFits (the code below does support different fits), all shapes (aspect ratio + circle) and rotations (which I can't get to work atm). 

My original thought with adding a listener won't work as you need way more details than just the movement data (screensize, cropsize, ...).

So I see two options. If you just need the "standard" (scaling/translating) case to work, you can fork the plugin and add the code below. Then you can use this fork in your project. I did just push it to a separate branch aswel, but this might be removed if it gets to far behind. [link](https://github.com/icapps/flutter-custom-image-crop/tree/feature/dont-go-out-of-bounds)
But if you like a challenge, feel free to continue where I left of and try to get rotation working. Bonus points for aspect ratio and circles. 

I hope this helps. 

```dart
  void onScaleUpdate(ScaleEvent event) {
    final scale = widget.canScale ? event.scale : (_dataTransitionStart?.scale ?? 1.0);
    final angle = widget.canRotate ? event.rotationAngle : 0.0;

    if (_dataTransitionStart != null) {
      if (!_canMove(
        data +
            (_dataTransitionStart! -
                CropImageData(
                  scale: scale,
                  angle: angle,
                )),
      )) return;

      addTransition(
        _dataTransitionStart! -
            CropImageData(
              scale: scale,
              angle: angle,
            ),
      );
    }
    _dataTransitionStart = CropImageData(
      scale: scale,
      angle: angle,
    );
  }

  void onMoveUpdate(MoveEvent event) {
    if (!widget.canMove) return;
    if (!_canMove(data + CropImageData(x: event.delta.dx, y: event.delta.dy))) return;

    addTransition(CropImageData(x: event.delta.dx, y: event.delta.dy));
  }

  bool _canMove(CropImageData data) {
    final image = _imageAsUIImage!;
    final cropFitParams = calculateCropFitParams(
      cropPercentage: widget.cropPercentage,
      imageFit: widget.imageFit,
      imageHeight: image.height,
      imageWidth: image.width,
      screenHeight: _height,
      screenWidth: _width,
      aspectRatio: (widget.ratio?.width ?? 1) / (widget.ratio?.height ?? 1),
    );
    final scale = data.scale * cropFitParams.additionalScale;
    final left = data.x - image.width / 2 * scale + cropFitParams.cropSizeWidth / 2;
    final right = data.x + image.width / 2 * scale - cropFitParams.cropSizeWidth / 2;
    final top = data.y - image.height / 2 * scale + cropFitParams.cropSizeHeight / 2;
    final bottom = data.y + image.height / 2 * scale - cropFitParams.cropSizeHeight / 2;
    return left < 0 && right > 0 && top < 0 && bottom > 0;
  }
```

**Comment by YumengNevix** [comment]
all gesture don't expose transition to other listener:
```dart
void onMoveUpdate(MoveEvent event) {
    if (!widget.canMove) return;
    /// this should be widget.controller.addTransition(xxxx)
    addTransition(CropImageData(x: event.delta.dx, y: event.delta.dy));
  }
```

Also, I think CustomImageCrop widget should add callback to expose Path rect and initial image rect like this:
<img width="484" alt="image" src="https://github.com/icapps/flutter-custom-image-crop/assets/137131451/49015eef-d9f4-4464-aa03-e127b46eab73">
<img width="406" alt="image" src="https://github.com/icapps/flutter-custom-image-crop/assets/137131451/5ceddbdf-1b62-4cfd-b396-8a50406aa66e">

If so, I can judge if image shows out of crop area and correct transition. 
<img width="643" alt="image" src="https://github.com/icapps/flutter-custom-image-crop/assets/137131451/c717ae8e-c112-466b-ade9-8a1a79ddb8e1">

**Comment by YumengNevix** [comment]
> all gesture don't expose transition to other listener:
> 
> ```dart
> void onMoveUpdate(MoveEvent event) {
>     if (!widget.canMove) return;
>     /// this should be widget.controller.addTransition(xxxx)
>     addTransition(CropImageData(x: event.delta.dx, y: event.delta.dy));
>   }
> ```
> 
> Also, I think CustomImageCrop widget should add callback to expose Path rect and initial image rect like this: <img alt="image" width="484" src="https://user-images.githubusercontent.com/137131451/268440366-49015eef-d9f4-4464-aa03-e127b46eab73.png"> <img alt="image" width="406" src="https://user-images.githubusercontent.com/137131451/268440485-5ceddbdf-1b62-4cfd-b396-8a50406aa66e.png">
> 
> If so, I can judge if image shows out of crop area and correct transition. <img alt="image" width="643" src="https://user-images.githubusercontent.com/137131451/268440512-c717ae8e-c112-466b-ade9-8a1a79ddb8e1.png">

also , I suggest add minScale and maxScale param, not all apps use 0.1-10 range -.-

**Comment by YumengNevix** [comment]
> @vlkonoshenko I've tried, but I can't get it to work 😅 fully. 
> 
> What you'll need to do is add a check in the `onMoveUpdate` and `onScaleUpdate`. Then in the check you'll need to check whether the image is in the bounds. This is the part where I'm having trouble, especially with the rotation. 
> This is some code that kind of works (movement and scaling is working). 
> But I can't add this to the package unless it works for all customImageFits (the code below does support different fits), all shapes (aspect ratio + circle) and rotations (which I can't get to work atm). 
> 
> My original thought with adding a listener won't work as you need way more details than just the movement data (screensize, cropsize, ...).
> 
> So I see two options. If you just need the "standard" (scaling/translating) case to work, you can fork the plugin and add the code below. Then you can use this fork in your project. I did just push it to a separate branch aswel, but this might be removed if it gets to far behind. [link](https://github.com/icapps/flutter-custom-image-crop/tree/feature/dont-go-out-of-bounds)
> But if you like a challenge, feel free to continue where I left of and try to get rotation working. Bonus points for aspect ratio and circles. 
> 
> I hope this helps. 
> 
> ```dart
>   void onScaleUpdate(ScaleEvent event) {
>     final scale = widget.canScale ? event.scale : (_dataTransitionStart?.scale ?? 1.0);
>     final angle = widget.canRotate ? event.rotationAngle : 0.0;
> 
>     if (_dataTransitionStart != null) {
>       if (!_canMove(
>         data +
>             (_dataTransitionStart! -
>                 CropImageData(
>                   scale: scale,
>                   angle: angle,
>                 )),
>       )) return;
> 
>       addTransition(
>         _dataTransitionStart! -
>             CropImageData(
>               scale: scale,
>               angle: angle,
>             ),
>       );
>     }
>     _dataTransitionStart = CropImageData(
>       scale: scale,
>       angle: angle,
>     );
>   }
> 
>   void onMoveUpdate(MoveEvent event) {
>     if (!widget.canMove) return;
>     if (!_canMove(data + CropImageData(x: event.delta.dx, y: event.delta.dy))) return;
> 
>     addTransition(CropImageData(x: event.delta.dx, y: event.delta.dy));
>   }
> 
>   bool _canMove(CropImageData data) {
>     final image = _imageAsUIImage!;
>     final cropFitParams = calculateCropFitParams(
>       cropPercentage: widget.cropPercentage,
>       imageFit: widget.imageFit,
>       imageHeight: image.height,
>       imageWidth: image.width,
>       screenHeight: _height,
>       screenWidth: _width,
>       aspectRatio: (widget.ratio?.width ?? 1) / (widget.ratio?.height ?? 1),
>     );
>     final scale = data.scale * cropFitParams.additionalScale;
>     final left = data.x - image.width / 2 * scale + cropFitParams.cropSizeWidth / 2;
>     final right = data.x + image.width / 2 * scale - cropFitParams.cropSizeWidth / 2;
>     final top = data.y - image.height / 2 * scale + cropFitParams.cropSizeHeight / 2;
>     final bottom = data.y + image.height / 2 * scale - cropFitParams.cropSizeHeight / 2;
>     return left < 0 && right > 0 && top < 0 && bottom > 0;
>   }
> ```

This method has bug, for example, when you move 30 pixels up but only 20 pixels left to reach the edge, image will not move, but the correct way is to move 20 pixels up, right?

**Comment by Muzammil-Bit** [comment]
Hi there, Is it possible to recenter the image whenever it goes out of the bounds of the viewport?

**Comment by ikbendewilliam** [comment]
@YumengNevix you are correct, but I didn't want to go that far. I saw your PR and will check it out, thank you for this.

@Muzammil-Bit The issue is that this is not so easy to detect with all the possible transformations. It might be that with the PR from @YumengNevix it's fixed (or at least good enough to add to the project).

**Comment by ikbendewilliam** [comment]
Thanks to @YumengNevix, this has now been fixed, check out version 0.0.13 and enable `forceInsideCropArea` to enable it.
